File: phutball/game/game_logic/jump.py
from .direction      import directions
from .location_state import empty, player, ball

class Jump:
  '''Class to represent and calculate jumps'''

  def __init__(self, path, end_state):
    '''Instantiate with a singleton or list of intermediate/final
    locations for the ball, and the resulting the state'''
    self._path      = path
    self.new_state  = end_state

  @property
  def path(self):
    if isinstance(self._path, list):
      return self._path
    else:
      return [self._path]
  
  def __repr__(self):
    loc_strings = map(str, self.path)
    return '-'.join(loc_strings)

  def __eq__(self, jump):
    if isinstance(jump, str):
      return str(self) == jump
    else:
      return str(self) == str(jump)

  def __add__(self, other):
    '''Add two jumps, with the temporally/causally
    earlier jump on the *right*'''

    new_path  = self.path + other.path
    new_state = other.new_state
    return Jump(new_path, new_state)

  @classmethod
  def get_legal_jumps(self, board):
    '''Returns a list of allowed jumps'''
    legal_jumps = (self._get_legal_jumps_from_rest(board, direction) for direction in directions)
    return sum(legal_jumps, [])
  
  @classmethod
  def _get_legal_jumps_from_rest(self, board, direction):
    target_loc = board.ball_loc + direction

    # Case 1: Illegal destination
    if not target_loc or not target_loc.on_board:
      return []

    # Case 2: Nothing to jump over
    elif board[target_loc] is empty:
      return []

    # Case 3: Something to jump over
    elif board[target_loc] is player:
      new_state = board.copy()
      new_state._move_ball(target_loc)
      return self._get_legal_jumps_from_motion(new_state, direction)

  @classmethod
  def _get_legal_jumps_from_motion(self, board, direction):
    target_loc = board.ball_loc + direction

    # Case 1: Tried to jump to nowhere
    if not target_loc:
      return []

    # Advance the piece
    previous_occupant = board[target_loc]
    new_state = board.copy()
    new_state._move_ball(target_loc)

    # Case 2: Target is occupied: Jump must continue
    if previous_occupant is player:
      return self._get_legal_jumps_from_motion(new_state, direction)

    # Case 3: Finish the jump.
    elif previous_occupant is empty:
      jump = Jump(target_loc, new_state)
      out  = [jump]
      for future_jump in self.get_legal_jumps(new_state):
        out.append(jump + future_jump)
      return out

File: phutball/frontend/src/gameLogic/jump.js
import directions from './direction'
import { empty, player } from './locationState'
import Tree from '../structs/tree'

// Wraps the jumps as a tree
export default function getLegalJumps(boardState) {
	const jumpList = Jump.getLegalJumps(boardState)	
	return Tree.buildTreeAsChildofMatch(jumpList, predecessorPredicate)
}


function predecessorPredicate(child, parent) {
	return parent.isPredecessorOf(child)
}

class Jump {
	constructor(path, endState, removedLocations, predecessor) {
		this._path            = path
		this.endState         = endState
		this.removedLocations = removedLocations
	}

	get path() {
		if (Array.isArray(this._path)) {
			return this._path
		} else {
			return [this._path]
		}
	}

	get endLoc() {
		return this.path[this.path.length-1]
	}

	toString() {
		return '*'+this.path.map(loc => loc.toString()).join('-')
	}

	isPredecessorOf(other) {
		if (other.path.length <= this.path.length) {
			return false
		}
		for (var step = 0 ; step < this.path.length; step++) {
			if (other.path[step] != this.path[step]) {
				return false
			}
		}
		return true
	}

	prependTo(nextJump) {
		const newPath          = this.path.concat(nextJump.path)
		const removedLocations = this.removedLocations.concat(nextJump.removedLocations)

		return new Jump(newPath, nextJump.endState, removedLocations)
	}

	static getLegalJumps(boardState) {
		const options  = directions.map(direction => Jump._fromRest(boardState, direction))

		// Join (concat) and return
		return options.reduce((left, right) => left.concat(right))
	}

	static _fromRest(boardState, direction) {
		const targetLoc = direction.add(boardState.ballLoc)

		// Don't attempt to jump to a place that does not exist
		if (targetLoc === null) {
			return []
		}

		// No jumping immediately to the off-board goalline from rest
		else if (!targetLoc.onBoard) {
			return []
		}

		// No jumping if there is nothing there to jump over
		else if (!(boardState.getSpace(targetLoc) === player)) {
			return []
		}

		// No jumping from the off-board goal-line
		else if (!boardState.ballLoc.onBoard) {
			return []
		}

		else {
			var intermediateState = boardState.copy()
			intermediateState._moveBall(targetLoc)

			const removedLocations = [targetLoc]

			return Jump._fromMotion(intermediateState, direction, removedLocations)
		}
	}

	static _fromMotion(boardState, direction, removedLocations) {
		const targetLoc = direction.add(boardState.ballLoc)

		// Don't attempt to jump to a place that does not exist
		if (targetLoc === null) {
			return []
		}

		// Advance the piece
		const previousOccupant = boardState.getSpace(targetLoc)
		var nextState = boardState.copy()
		nextState._moveBall(targetLoc)

		// If jumping to the goalline or the target is empty, land!
		//  And then consider more jumps
		if (!(targetLoc.onBoard) || (previousOccupant === empty)) {

			const jump = new Jump(targetLoc, nextState, removedLocations);

			var out = [jump];
			Jump.getLegalJumps(nextState).forEach(nextJump => out.push(jump.prependTo(nextJump)))
			return out
		}

		// If the target is occupied, the jump must continue
		else if (previousOccupant === player) {
			removedLocations.push(targetLoc)
			return Jump._fromMotion(nextState, direction, removedLocations)
		} 

	}

};
